### ISSUE TITLE
current raid/mdadm support does not allow enough time for slow devices to come fully online

### ISSUE THREAD

**Issue by richard-cms** [anatol/booster#97]
I'm using `booster 0.6-1` from the archlinux repositories. My root is a 2-disk RAID1 on relatively slow HDDs. I am not able to use booster to boot because I believe that booster is not waiting for the RAID1 to fully finish assembly before attempting to mount. My system does alternatively work with `mkinitcpio` and the `systemd` + `mdadm_udev` hooks.

When attempting to boot, I get an "unable to read superblock" error. If I wait for the `mount_timeout` to elapse I can manually mount the RAID device, which is already present, without error.

An abbreviated dmesg log is here (I can attach full log if requested). Notice that the "unable to read superblock" error occurs moments before the raid device comes fully online:
```
[    0.000000] Command line: initrd=\amd-ucode.img initrd=\booster-linux-lts.img root=/dev/md/root rootfstype=ext4 rw
[    0.000000] Kernel command line: initrd=\amd-ucode.img initrd=\booster-linux-lts.img root=/dev/md/root rootfstype=ext4 rw
[    0.130695] smpboot: CPU0: AMD Ryzen 3 2200G with Radeon Vega Graphics (family: 0x17, model: 0x11, stepping: 0x0)
[    6.369652] md_mod: module verification failed: signature and/or required key missing - tainting kernel
[    6.462668] EXT4-fs (md127): unable to read superblock
[    6.462755] booster: mount(/dev/md/root): input/output error
[    6.505324] md/raid1:md127: active with 2 out of 2 mirrors
[    6.543912] md127: detected capacity change from 0 to 999546290176
[  128.412919] EXT4-fs (md127): mounted filesystem with ordered data mode. Opts: (null)
```
(I'm also very surprised by the `tainting kernel` line but that seems unrelated)

My `/etc/booster.yaml` is
```
universal: false
compression: zstd
mount_timeout: 1m
strip: true
# I removed fsck and fsck.ext4 in the course of debugging
extra_files: busybox
enable_mdraid: true
```

I'm using `systemd-boot` and my boot entry is:
```
title Arch Linux
linux /vmlinuz-linux-lts
initrd /amd-ucode.img
initrd /booster-linux-lts.img
options root=/dev/md/root rootfstype=ext4 rw
```

My  `/etc/mdadm.conf` is just
```
DEVICE partitions
ARRAY /dev/md/root metadata=1.2 name=<hostname>:root UUID=<UUID>
```

**Comment by anatol** [comment]
Thank you for the feedback.

```
[    6.462668] EXT4-fs (md127): unable to read superblock
[    6.462755] booster: mount(/dev/md/root): input/output error
[    6.505324] md/raid1:md127: active with 2 out of 2 mirrors
[    6.543912] md127: detected capacity change from 0 to 999546290176
```

It is an interesting behavior. I would expect that `mdadm --incremental` leaves raid in a working condition. i.e. once `mdadm` completes and udev event is fired then one can use the block device. But as you pointed out it might not be the case.

Could you please enable booster debugging with `booster.debug` kernel param and then post your output? It will help to understand the mdadm behavior better.

BTW here are the udev rules from Arch for the same even:

```
ACTION=="add|change", IMPORT{program}="/usr/bin/mdadm --incremental --export $devnode --offroot $env{DEVLINKS}"
ACTION=="add|change", ENV{MD_STARTED}=="*unsafe*", ENV{MD_FOREIGN}=="no", ENV{SYSTEMD_WANTS}+="mdadm-last-resort@$env{MD_DEVICE}.timer"
```

It uses `--incremental` as well (the same as booster). It also uses `--offroot` flag that is not part of manpage, interesting.... And then it can call `mdadm-last-resort` which is labeled as "Timer to wait for more drives before activating degraded array %I.". I will need to read more about it to understand how mdadm handles it.

But other than that I do not see any delays or extra sync functionality. I wonder if booster is just "too fast" and starts mounting the block device right away, while udev has a lot of delays internally and hence we do not see this problem there.

```
[    6.369652] md_mod: module verification failed: signature and/or required key missing - tainting kernel
```
As of this warning - could you please file a ticket and we proceed its discussion separately.

**Comment by anatol** [comment]
Also could you please share configuration for your mdraid array so I'll try to reproduce the issue it at my side.

**Comment by anatol** [comment]
```
[    6.462668] EXT4-fs (md127): unable to read superblock
[    6.462755] booster: mount(/dev/md/root): input/output error
[    6.505324] md/raid1:md127: active with 2 out of 2 mirrors
[    6.543912] md127: detected capacity change from 0 to 999546290176
```

Returning back to this log. It looks like booster gets a udev event for `/dev/md/root` and assumes the device is ready. While it is not. Only 38 ms later kernel reports the array is assembled. We need to figure out under what conditions such "early" udev events are reported. Having `booster.debug` logs might help to shed light on it.

Also locally I added a RAID1 integration test with 5 partitions and it works fine. The array is fully assembled first and then kernel sends a udev event that is handled by booster correctly.

**Comment by richard-cms** [comment]
> Was you able to boot successfully without* debug enabled as well?

No! I confirmed a few times that I could *not* boot unless the `booster.debug` flag was present. 

There is no visible error that shows up now, just the timeout waiting for root. The dmesg log shows that `/dev/md127` does eventually get assembled but no actions are taken after it is.

**Comment by anatol** [comment]
> No! I confirmed a few times that I could not boot unless the booster.debug flag was present.

It is interesting... It also gives an opportunity to understand what is going on there.

I just pushed a commit to `wip` branch to print mdadm output to stdout. If you have a chance could you please compile booster from `wip` branch and run it without `booster.debug`. There should be extra information at the console. Please capture it and post it here.

**Comment by richard-cms** [comment]
I rebuilt booster `init` from the `wip` branch and the following additional lines are printed during boot (`debug.booster` is *not* set). Boot is not successful, it hangs after these messages. These are transcribed by hand from screen, sorry if there are minor typos.

```
&{Header:add@/devices/virtual/block/md127 Action:add Devpath:/devices/virtual/block/md127 Subsystem:block Seqnum:1399 Vars:map[ACTION:add DEVNAME:md127 DEVPATH:/devices/virtual/block/md127 DEVTYPE:disk MAJOR:9 MINOR:127 SEQNUM:1299 SUBSYSTEM:block]}

mdadm : /dev/sda2 attached to /dev/md/root, not enough to start safely.
&{Header:change@/devices/virtual/block/md127 Action:change Devpath:/devices/virtual/block/md127 Subsystem:block Seqnum:1399 Vars:map[ACTION:change DEVNAME:md127 DEVPATH:/devices/virtual/block/md127 DEVTYPE:disk MAJOR:9 MINOR:127 SEQNUM:1299 SUBSYSTEM:block]}

mdadm:  /dev/sdb1 attached to /dev/md/root, which has been started.
```

**Comment by anatol** [comment]
Thank you for the logs. It looks like there is a short period of time when the raid1 array exists but not fully assembled yet. Instead of relying on udev "add" event booster should use `mdadm --export` and check whether mdadm was able to start the array.

Here is a [proposed fix](https://github.com/anatol/booster/commit/4e736d42ee2891572d33a5bff3b5c2ca1d30abb8) for the problem that has been pushed to `wip` branch. @richard-cms please give it a try and let me know it fixes your problem (both with debug enabled and disabled).

**Comment by richard-cms** [comment]
I have rebuilt based on 3fd94442695196fdd873eb205644dad28598804d (HEAD of `wip` as of 2021-08-29 EST afternoon) and I now successfully boot both with and without `booster.debug` set. I rebooted a few times, including 'cold boots', to see if there were any common race conditions and did not run in to any. 

Attached is the `dmesg` log for a successful boot with `booster.debug` set. Thanks very much @anatol ! 
[2021-08-29-booster_debug_dmesg.txt](https://github.com/anatol/booster/files/7073211/2021-08-29-booster_debug_dmesg.txt)
